File: js/src/tools/audit.ts
import { Tool, ToolContext, ToolResult } from "./base.js";
import { getSandbox } from "../computer/sandbox.js";

// Actions that produce no side-effects — skip during replay.
const SKIP_REPLAY = new Set([
  "cursor_position", "screenshot", "ocr", "clipboard_read",
  "ui_get_tree", "ui_get_value", "app_list",
]);

// Maps audit action name → tool name for dispatch.
const ACTION_TO_TOOL: Record<string, string> = {
  keyboard_type: "keyboard", keyboard_press: "keyboard",
  keyboard_hotkey: "keyboard", keyboard_hold: "keyboard",
  mouse_click: "mouse", mouse_move: "mouse", mouse_double_click: "mouse",
  mouse_right_click: "mouse", mouse_scroll: "mouse", mouse_drag: "mouse",
  app_open: "app", app_close: "app", app_focus: "app",
  clipboard_write: "clipboard",
  ui_click: "ui", ui_click_menu: "ui", ui_type: "ui", ui_press_key: "ui",
};

async function loadTool(name: string): Promise<Tool> {
  switch (name) {
    case "keyboard": return new (await import("./keyboard.js")).KeyboardTool();
    case "mouse":    return new (await import("./mouse.js")).MouseTool();
    case "app":      return new (await import("./app.js")).AppTool();
    case "clipboard":return new (await import("./clipboard.js")).ClipboardTool();
    case "ui":       return new (await import("./ui.js")).UITool();
    default: throw new Error(`Unknown tool: ${name}`);
  }
}

export class AuditTool extends Tool {
  name = "audit";
  description =
    "Show or replay the session audit log.\n" +
    "  action='show'   — display recorded actions (format='summary' or 'full')\n" +
    "  action='replay' — re-execute every action from this session " +
    "(skips screenshots, reads, and optionally failed actions)";

  schema = {
    type: "object",
    properties: {
      action: { type: "string", enum: ["show", "replay"], default: "show" },
      format: { type: "string", enum: ["summary", "full"], default: "full",
                description: "For action=show: summary or full timestamped log" },
      sessionId: { type: "string", description: "Session to inspect — defaults to current session" },
      skipErrors: { type: "boolean", default: true,
                    description: "For action=replay: skip entries that originally errored" },
    },
  };

  async execute(ctx: ToolContext, params: Record<string, unknown>): Promise<ToolResult> {
    const action = (params.action as string) ?? "show";
    const sessionId = (params.sessionId as string) ?? ctx.sessionId;
    const sandbox = getSandbox(sessionId);

    if (action === "replay") {
      const skipErrors = (params.skipErrors as boolean) ?? true;
      return this._replay(ctx, sandbox, skipErrors);
    }

    // --- show ---
    const format = (params.format as string) ?? "full";
    if (format === "summary") {
      return this.ok("Audit summary", sandbox.summary());
    }

    const log = sandbox.exportAuditLog();
    if (!log.length) {
      return this.ok("Audit log", `No actions recorded yet for session '${sessionId}'.`);
    }

    const lines = [`Audit log — session '${sessionId}' (${log.length} actions)\n`];
    for (let i = 0; i < log.length; i++) {
      const entry = log[i];
      const ts = new Date(entry.timestamp * 1000).toISOString().replace("T", " ").slice(0, 19);
      const errorTag = entry.error ? `  ERROR: ${entry.error}` : "";
      const replayable = entry.replayParams || ACTION_TO_TOOL[entry.action] ? "✓" : " ";
      lines.push(
        `[${String(i + 1).padStart(3)}] [${replayable}] [${ts}] ${entry.action.padEnd(20)} ${JSON.stringify(entry.params)}${errorTag}`
      );
    }
    lines.push("\n✓ = replayable via audit(action='replay')");
    return this.ok("Audit log", lines.join("\n"));
  }

  private async _replay(
    ctx: ToolContext,
    sandbox: ReturnType<typeof getSandbox>,
    skipErrors: boolean,
  ): Promise<ToolResult> {
    const log = sandbox.exportAuditLog();
    if (!log.length) {
      return this.ok("Audit replay", "No actions to replay — the audit log is empty.");
    }

    const replayable = log.filter(
      (e) =>
        !SKIP_REPLAY.has(e.action) &&
        ACTION_TO_TOOL[e.action] !== undefined &&
        !(skipErrors && e.error),
    );

    if (!replayable.length) {
      return this.ok("Audit replay", "No replayable actions found in this session.");
    }

    const lines = [`Replaying ${replayable.length} action(s)...\n`];
    let ok = 0;
    let failed = 0;

    for (const entry of replayable) {
      const toolName = ACTION_TO_TOOL[entry.action];
      // Use replayParams if present (ui actions), otherwise params is already full (keyboard/mouse/etc.)
      const replayParams = entry.replayParams ?? entry.params;

      try {
        const tool = await loadTool(toolName);
        const result = await tool.execute(ctx, replayParams);
        const status = result.error ? "ERROR" : "OK";
        lines.push(`  ${status.padEnd(5)} ${entry.action.padEnd(20)} ${result.output.slice(0, 80)}`);
        if (result.error) failed++; else ok++;
      } catch (e) {
        lines.push(`  ERROR ${entry.action.padEnd(20)} ${e}`);
        failed++;
      }
    }

    lines.push(`\nDone — ${ok} succeeded, ${failed} failed.`);
    return this.ok("Audit replay", lines.join("\n"));
  }
}

File: python/opendesk/tools/audit.py
"""Audit tool — expose the session audit log inside any MCP or agent session."""

from __future__ import annotations

import json
from typing import Literal, Optional

from pydantic import Field

from opendesk.tools.base import Tool, ToolContext, ToolResult


# ActionTypes that produce no side-effects and should be skipped during replay.
_SKIP_REPLAY = frozenset({
    "screenshot", "cursor_position", "app_list", "clipboard_read", "ocr",
})

# Maps audit action value → tool name for lazy import during replay.
_ACTION_TO_TOOL: dict[str, str] = {
    "mouse_click": "mouse", "mouse_move": "mouse", "mouse_scroll": "mouse",
    "mouse_drag": "mouse", "mouse_down": "mouse", "mouse_up": "mouse",
    "keyboard_type": "keyboard", "keyboard_press": "keyboard",
    "keyboard_hotkey": "keyboard", "keyboard_hold": "keyboard",
    "app_open": "app", "app_close": "app", "app_focus": "app",
    "clipboard_write": "clipboard",
    "ui_action": "ui",
}

_TOOL_CLASSES = {
    "mouse": "opendesk.tools.mouse:MouseTool",
    "keyboard": "opendesk.tools.keyboard:KeyboardTool",
    "app": "opendesk.tools.app:AppTool",
    "clipboard": "opendesk.tools.clipboard:ClipboardTool",
    "ui": "opendesk.tools.ui:UITool",
}


def _load_tool(name: str):
    import importlib
    module_path, cls_name = _TOOL_CLASSES[name].split(":")
    mod = importlib.import_module(module_path)
    return getattr(mod, cls_name)()


class AuditTool(Tool):
    """Show or replay the session audit log.

    action='show'   — display the log (summary or full timestamped list).
    action='replay' — re-execute every replayable action recorded this session.
    """

    name = "audit"
    description = (
        "Show or replay the session audit log.\n"
        "  action='show'   — display recorded actions (format='summary' or 'full')\n"
        "  action='replay' — re-execute every action from this session "
        "(skips screenshots, reads, and optionally failed actions)"
    )

    class Params(Tool.Params):
        action: Literal["show", "replay"] = Field(
            default="show",
            description="'show' displays the log; 'replay' re-executes actions.",
        )
        format: Literal["summary", "full"] = Field(
            default="full",
            description="For action='show': 'summary' = one-line count, 'full' = timestamped log.",
        )
        session_id: Optional[str] = Field(
            default=None,
            description="Session to inspect/replay. Defaults to the current session.",
        )
        skip_errors: bool = Field(
            default=True,
            description="For action='replay': skip actions that originally errored.",
        )

    async def execute(self, ctx: ToolContext, params: "AuditTool.Params") -> ToolResult:
        from opendesk.computer.sandbox import get_sandbox

        session_id = params.session_id or ctx.session_id
        sandbox = get_sandbox(session_id)

        if params.action == "replay":
            return await self._replay(ctx, sandbox, params.skip_errors)

        # --- show ---
        if params.format == "summary":
            return ToolResult(title="Audit summary", output=sandbox.summary())

        log = sandbox.export_audit_log()
        if not log:
            return ToolResult(
                title="Audit log",
                output=f"No actions recorded yet for session '{session_id}'.",
            )

        import datetime
        lines: list[str] = [f"Audit log — session '{session_id}' ({len(log)} actions)\n"]
        for i, entry in enumerate(log, 1):
            ts = datetime.datetime.fromtimestamp(entry["timestamp"]).strftime(
                "%Y-%m-%d %H:%M:%S"
            )
            error_tag = f"  ERROR: {entry['error']}" if entry["error"] else ""
            replayable = "✓" if entry.get("replay_params") else " "
            params_str = json.dumps(entry["params"], ensure_ascii=False)
            lines.append(
                f"[{i:>3}] [{replayable}] [{ts}] {entry['action']:<20} {params_str}{error_tag}"
            )

        lines.append("\n✓ = replayable via audit(action='replay')")
        return ToolResult(title="Audit log", output="\n".join(lines))

    async def _replay(
        self, ctx: ToolContext, sandbox, skip_errors: bool
    ) -> ToolResult:
        log = sandbox.export_audit_log()
        if not log:
            return ToolResult(
                title="Audit replay",
                output="No actions to replay — the audit log is empty.",
            )

        replayable = [
            e for e in log
            if e["action"] not in _SKIP_REPLAY
            and e.get("replay_params") is not None
            and (not skip_errors or not e["error"])
        ]

        if not replayable:
            return ToolResult(
                title="Audit replay",
                output=(
                    "No replayable actions found. "
                    "Replayable actions require replay_params — only actions recorded "
                    "in this session (not imported logs) support direct replay."
                ),
            )

        results: list[str] = [f"Replaying {len(replayable)} action(s)...\n"]
        ok = 0
        failed = 0

        for entry in replayable:
            action_val = entry["action"]
            tool_name = _ACTION_TO_TOOL.get(action_val)
            if tool_name is None:
                results.append(f"  SKIP  {action_val} (no tool mapping)")
                continue

            tool = _load_tool(tool_name)
            rp = entry["replay_params"]
            try:
                result = await tool.execute(ctx, tool.parse_params(rp))
                status = "ERROR" if result.error else "OK"
                results.append(f"  {status:<5} {action_val:<20} {result.output[:80]}")
                if result.error:
                    failed += 1
                else:
                    ok += 1
            except Exception as exc:
                results.append(f"  ERROR {action_val:<20} {exc}")
                failed += 1

        results.append(f"\nDone — {ok} succeeded, {failed} failed.")
        return ToolResult(title="Audit replay", output="\n".join(results))
